### PR TITLE
Enable time zone for netstandard1.7

### PR DESCRIPTION
**Pull request by tarekgh**
The change is exposing all TimeZone APIs from corelib so can be used in netstandard 1.7.
The change here is re-implementing CurrentTimeZone class to be simpler and also to make it work for Linux platforms. we used TimeZoneInfo class to re-implement it and also cleaned up all used internal calls as not needed anymore.
The change also fixes how we return the adjustment rules in TimeZoneInfo as it was returning not user friendly contents.
The changes here will make TimeZone behave better than the old implementation as before we used to work with rule matching current year only. I have to fix that because to make it work on Linux I didn’t have a way to keep the old wrong behavior.

### CODE REVIEW

**Comment by tarekgh** [comment]
https://github.com/dotnet/corefx/issues/11169

CC @jkotas @danmosemsft @weshaggard

**Comment by tarekgh** [comment]
by the way, TimeZone is already obsolete class but we had to expose it as the rest of the APIs in netstandard 2.

File: src/mscorlib/src/System/CurrentTimeZone.cs [hunk 1]
```diff
@@ -60,43 +45,40 @@ internal class CurrentSystemTimeZone : TimeZone {
...
-            [System.Security.SecuritySafeCritical]  // auto-generated
-            get {
-                if (m_standardName == null) {
-                    m_standardName = nativeGetStandardName();
-                }
+        public override String StandardName
+        {
+            get 
+            {
                 return (m_standardName);
```
Comment (danmosemsft) [inline src/mscorlib/src/System/CurrentTimeZone.cs]: parentheses?

File: src/mscorlib/src/System/CurrentTimeZone.cs [hunk 2]
```diff
@@ -107,119 +89,124 @@ internal class CurrentSystemTimeZone : TimeZone {
...
+            if (year < 1 || year > 9999)
+            {
                 throw new ArgumentOutOfRangeException(nameof(year), Environment.GetResourceString("ArgumentOutOfRange_Range", 1, 9999));
             }
             Contract.EndContractBlock();
 
-            Object objYear = (Object)year;
-
-            if (!m_CachedDaylightChanges.Contains(objYear)) {
-                lock (InternalSyncObject) {
```
Comment (danmosemsft) [inline src/mscorlib/src/System/CurrentTimeZone.cs]: did you intentionally remove this lock

File: src/mscorlib/src/System/TimeZoneInfo.cs [hunk 3]
```diff
@@ -319,24 +319,58 @@ private class OffsetAndRule {
...
+        // and can use it as it is used in Windows
 
-        // ---- SECTION: public methods --------------*
+        private AdjustmentRule [] GetFilledRules()
+        {
+            Contract.Assert(m_adjustmentRules != null, "m_adjustmentRules expected to be not null");
+            AdjustmentRule [] rules = new AdjustmentRule[m_adjustmentRules.Length];
+
+            for (int i=0; i<m_adjustmentRules.Length; i++)
+            {
```
Comment (danmosemsft) [inline src/mscorlib/src/System/TimeZoneInfo.cs]: nit.. spacings

File: src/mscorlib/src/System/TimeZoneInfo.cs [hunk 4]
```diff
@@ -319,24 +319,58 @@ private class OffsetAndRule {
...
+#if PLATFORM_UNIX
+        // The rules we use in Unix cares mostly about the start and end dates but doesn’t fill the transition start and end info. 
+        // as the rules now is public, we should fill it properly so the caller doesn’t have to know how we use it internally 
+        // and can use it as it is used in Windows
 
-        // ---- SECTION: public methods --------------*
+        private AdjustmentRule [] GetFilledRules()
+        {
+            Contract.Assert(m_adjustmentRules != null, "m_adjustmentRules expected to be not null");
+            AdjustmentRule [] rules = new AdjustmentRule[m_adjustmentRules.Length];
```
Comment (danmosemsft) [inline src/mscorlib/src/System/TimeZoneInfo.cs]: >e [ [](start = 25, length = 3)

nit..spacing

File: src/mscorlib/src/System/CurrentTimeZone.cs (hunk 2 above)
Comment (danmosemsft) [inline src/mscorlib/src/System/CurrentTimeZone.cs]: it seems it is protecting m_CachedDaylightChanges and you are still using it eselwhere

---
In reply to: [87494826](https://github.com/dotnet/coreclr/pull/8074#discussion_r87494826) [](ancestors = 87494826)

Comment (danmosemsft) [inline src/mscorlib/src/System/CurrentTimeZone.cs]: Hashtable is safe for one writer and multiple readers which may be why you removed this lock and only left the one around the .Add. Personally I just lock everywhere .. someone could change it to a dictionary later. Up to you. 

---
In reply to: [87495577](https://github.com/dotnet/coreclr/pull/8074#discussion_r87495577) [](ancestors = 87495577,87494826)

**Comment by danmosemsft** [comment]
reusing any Mono tests they have for TimeZone might help reassure us that your refactoring is good (I mean if you hadn't refactored we at least know it matches desktop which is good). Up to you

**Comment by danmosemsft** [comment]
:shipit:

File: src/mscorlib/src/System/CurrentTimeZone.cs (hunk 2 above)
Comment (danmosemsft) [inline src/mscorlib/src/System/CurrentTimeZone.cs]: I made another comment which seems to have gotten lost: you may be removed this lock as Hashtable is safe for single writer multiple reader. That's OK so long as nobody changes this to a Dictionary later without noticing you don't lock everywhere.

Comment (tarekgh) [inline src/mscorlib/src/System/CurrentTimeZone.cs]: I didn't remove the lock, I moved it down so the whole block we are creating a new item doesn't have inside the lock block. note that the code already reading from the hash table before locking too. so I didn't really change the logic

---
In reply to: [87496264](https://github.com/dotnet/coreclr/pull/8074#discussion_r87496264) [](ancestors = 87496264,87495577,87494826)

File: src/mscorlib/src/System/TimeZoneInfo.cs (hunk 4 above)
Comment (tarekgh) [inline src/mscorlib/src/System/TimeZoneInfo.cs]: just wondering why spacing here is not good?

---
In reply to: [87495205](https://github.com/dotnet/coreclr/pull/8074#discussion_r87495205) [](ancestors = 87495205)

**Comment by tarekgh** [comment]
I have created my own test in corefx side. just to clarify one issue I have mentioned in the issue description which is "The changes here will make TimeZone behave better than the old implementation as before we used to work with rule matching current year only. I have to fix that because to make it work on Linux I didn’t have a way to keep the old wrong behavior."


---
In reply to: [259822401](https://github.com/dotnet/coreclr/pull/8074#issuecomment-259822401) [](ancestors = 259822401)

File: src/mscorlib/src/System/TimeZoneInfo.cs (hunk 4 above)
Comment (danmosemsft) [inline src/mscorlib/src/System/TimeZoneInfo.cs]: I thought we don't put a space before the []. Maybe I'm wrong.

**Comment by tarekgh** [comment]
@jkotas could you please have a quick look as I have removed some of not needed internal calls? I want to make sure I am not missing anything, thanks,

**Comment by jkotas** [comment]
LGTM

**Comment by tarekgh** [comment]
@dotnet-bot test Windows_NT x64 Formatting please

**Comment by tarekgh** [comment]
@dotnet-bot test Windows_NT x64 Checked Jit Diff Build and Test please

**Pull request merged**